opencda/core/ml_libs/rl/utils/simulator_utils/sensor_utils.py

```diff
@@ -0,0 +1,565 @@
+import os
+import copy
+import logging
+import time
+import numpy as np
+import carla
+import math
+import weakref
+import shapely.geometry
+from enum import Enum
+from easydict import EasyDict
+from collections import deque
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from ding.utils.default_helper import deep_merge_dicts
+
+DEFAULT_CAMERA_CONFIG = {
+    'size': [384, 160],
+    'fov': 90,
+    'position': [2.0, 0.0, 1.4],
+    'rotation': [0, 0, 0],
+}
+
+DEFAULT_CAMERA_AUG_CONFIG = {'position_range': [0, 0, 0], 'rotation_range': [0, 0, 0]}
+
+DEFAULT_LIDAR_CONFIG = {
+    'channels': 1,
+    'range': 2000,
+    'points_per_second': 1000,
+    'rotation_frequency': 10,
+    'upper_fov': -3,
+    'lower_fov': -3,
+    'position': [0, 0.0, 1.4],
+    'rotation': [0, -90, 0],
+    'draw': False,
+}
+
+DEFAULT_GNSS_CONFIG = {
+    'position': [0.0, 0.0, 1.4],
+}
+
+
+class TrafficLightState(Enum):
+    RED = 0
+    YELLOW = 1
+    GREEN = 2
+    OFF = 3
+
+
+def get_random_sample(range_list):
+    res = []
+    for _range in range_list:
+        num = np.random.random() * _range * 2 - _range
+        res.append(num)
+    return res
+
+class SensorHelper(object):
+    """
+    Interfaces for sensors required for vehicles and data buffer for all sensor data in Carla. The updating for Carla
+    sensors are not synchronous. In each tick, the newest sensor data is obtained from sensor data buffer and returned
+    to the simulator. This class provides an interface that can easily create, receive data and destroy all
+    kinds of sensors in Carla according to config, and apply the same sensor augmantation to all camera sensors.
+
+    :Arguments:
+        - obs_cfg (Dict): Config dict for sensor
+        - aug_cfg (Dict, optional): Config dict for sensor augmentation. Defaults to None.
+
+    :Interfaces: setup_sensors, get_sensors_data, clear_up
+    """
+
+    def __init__(
+            self,
+            obs_cfg: Dict,
+            aug_cfg: Optional[Dict] = None,
+    ) -> None:
+        self._obs_cfg = obs_cfg
+        self._obs_cfg_list = [self._obs_cfg]
+        self._aug_cfg = aug_cfg
+        self._sensors_dict = {}
+        self._data_buffers = {}
+        self._timestamps = {}
+        self._random_aug_pos = None
+        self._random_aug_rot = None
+
+    def setup_sensors(self, world: carla.World, vehicle: carla.Actor) -> None:
```

Comment:
This class is overlapped with our own perception_manager. Please reuse our original implementation